File: apps/slate-mac/Sources/SlateMac/Graph/GraphAnnouncer.swift
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

import AppKit
import Foundation

// MARK: - Generated-enum extensions (W6-2 PR 0a; contracts doc 0a-5, 0a-18)

// `GraphVerbosity` is core's since W6-2 PR 0a (`slate_core::a11y`). This
// extension gives the generated enum what a store and a settings picker
// need — the canvas extension's shape (`CanvasAnnouncer.swift`). Nothing
// persists the level YET: mac never did (P2-4's intention never landed),
// and 0b adds the `verbosity` key to `.slate/graph.json`'s schema; until
// then every announcer starts at `.standard`.
//
// **Persistence-tag stability — DO NOT RENAME.** The literals in
// `persistenceTag` are the strings 0b's store writes into the vault
// file; renaming one silently resets a user's choice.
extension GraphVerbosity: Codable, CaseIterable {
    public static var allCases: [GraphVerbosity] {
        [.terse, .standard, .verbose]
    }

    /// Settings picker label (t0 §1.2 level names).
    var title: String {
        switch self {
        case .terse: return "Terse"
        case .standard: return "Standard"
        case .verbose: return "Verbose"
        }
    }

    private var persistenceTag: String {
        switch self {
        case .terse: return "terse"
        case .standard: return "standard"
        case .verbose: return "verbose"
        }
    }

    public init(from decoder: Decoder) throws {
        let value = try decoder.singleValueContainer().decode(String.self)
        switch value {
        case "terse": self = .terse
        case "standard": self = .standard
        case "verbose": self = .verbose
        default:
            throw DecodingError.dataCorruptedError(
                in: try decoder.singleValueContainer(),
                debugDescription: "Unknown GraphVerbosity: \(value)"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(persistenceTag)
    }
}

// The Graph tab's two projections are core's `GraphSurfaceMode` (W6-2 PR
// 0a) — the local `GraphTabMode` it replaces was a `String`-raw-valued
// enum, so the config store's `mode` key keeps the SAME two tags through
// this explicit mapping (P2-4 v1 schema, unchanged on disk).
//
// **Persistence-tag stability — DO NOT RENAME.**
extension GraphSurfaceMode: CaseIterable {
    public static var allCases: [GraphSurfaceMode] {
        [.table, .diagram]
    }

    var title: String {
        switch self {
        case .table: return "Table"
        case .diagram: return "Diagram"
        }
    }

    /// The `.slate/graph.json` `mode` tag.
    var persistenceTag: String {
        switch self {
        case .table: return "table"
        case .diagram: return "diagram"
        }
    }

    init?(persistenceTag: String) {
        switch persistenceTag {
        case "table": self = .table
        case "diagram": self = .diagram
        default: return nil
        }
    }
}

// MARK: - The announcer

/// The one announcement funnel for every graph surface (P1-1 #554; W6-2
/// PR 0a, #746).
///
/// Since W6-2 PR 0a the GRAMMAR is core's: callers hand over a typed
/// `GraphA11yEvent` from `slate_core::a11y` and this class renders it
/// through the FFI, so the spoken text AND the priority both come from
/// the vocabulary — never composed here, never re-classified here. What
/// stays host-side is exactly what a pure render cannot own: the ~200 ms
/// same-class coalescing window (a render has no clock), the filter
/// count's fire-time relevance gate, `cancelPending` on view departure,
/// and the live-switchable verbosity the row family takes as a parameter
/// (contracts doc 0a-9, 0aD-3). The class membership is pinned core-side
/// in the one "Coalescing class keys" list; slate-uniffi's
/// `the_mac_graph_coalescing_switch_matches_the_pinned_class_list` reads
/// the switch below against it.
///
/// No graph code calls `postAccessibilityAnnouncement` directly — DoD §H,
/// enforced by `GraphAnnouncerTests.testNoDirectAnnouncementsUnderGraph`.
@MainActor
final class GraphAnnouncer: ObservableObject {
    /// Live-switchable verbosity, passed per row event (core holds no
    /// "current verbosity"). Default-only until 0b's store lands.
    @Published var verbosity: GraphVerbosity

    /// Event classes for coalescing: same-class events within the window
    /// collapse to the latest; each class is independent, so a settled
    /// message can neither cancel nor be cancelled by a force value.
    private enum EventClass: Hashable {
        case navigation
        case filter
        case forceValue
        case settle
    }

    /// The production coalescing window (t0 §1.5's ~200 ms), test-visible
    /// so the pin is a number and not a stopwatch (contracts doc 0a-9).
    static let defaultCoalesceWindow: TimeInterval = 0.2

    private let post: (String, NSAccessibilityPriorityLevel) -> Void
    private let coalesceWindow: TimeInterval
    private var pending:
        [EventClass: (
            work: DispatchWorkItem, text: String, priority: NSAccessibilityPriorityLevel,
            gate: () -> Bool
        )] = [:]

    /// The seam is `(text, priority)` rather than an event because the
    /// coalescer holds a RENDERED line: the window's winner is decided
    /// after the render, and the loser is dropped without being spoken.
    /// The default posts through the poster layer — never the string
    /// primitive, never `.hostComposed` (the residue census dropped this
    /// class's site: 29 → 28).
    init(
        verbosity: GraphVerbosity = .standard,
        coalesceWindow: TimeInterval = GraphAnnouncer.defaultCoalesceWindow,
        post: @escaping (String, NSAccessibilityPriorityLevel) -> Void = { text, priority in
            let level: AnnouncementPriority = priority == .high ? .high : .medium
            AppKitAnnouncementPoster().post(text, priority: level)
        }
    ) {
        self.verbosity = verbosity
        self.coalesceWindow = coalesceWindow
        self.post = post
    }

    /// The only announcement API graph code may use.
    func announce(_ event: GraphA11yEvent) {
        emit(.graph(event: event), coalescing: Self.coalescingClass(of: event), gate: { true })
    }

    /// The ONE gated entry (contracts doc 0a-13, 0aD-3): the filter count
    /// carries a FIRE-TIME relevance gate — a count queued while the graph
    /// was focused is dropped if focus has since moved to another split
    /// pane (which leaves the graph mounted, so `onDisappear` never runs).
    /// Restricted to the filter class: no other event can acquire a host
    /// relevance gate.
    func announceFilterCount(shown: UInt32, total: UInt32, gate: @escaping () -> Bool) {
        let event: GraphA11yEvent = .graphFilterCount(shown: shown, total: total)
        emit(.graph(event: event), coalescing: Self.coalescingClass(of: event), gate: gate)
    }

    /// Relay a core event that is NOT graph vocabulary — the shared data
    /// grid raises its own sort/filter events, and the graph table must
    /// pass their rendered text AND priority through rather than
    /// re-wrapping them at a priority of its own choosing (0a-13, 0a-D2).
    func relay(_ event: A11yEvent) {
        emit(event, coalescing: nil, gate: { true })
    }

    /// The copy of core's class list (a11y.rs "Coalescing class keys",
    /// the graph rows). Parsed by slate-uniffi's tripwire: keep the shape.
    private static func coalescingClass(of event: GraphA11yEvent) -> EventClass? {
        switch event {
        case .graphRow: return .navigation
        case .graphFilterCount: return .filter
        case .graphForceValue: return .forceValue
        case .graphLayoutSettled: return .settle
        default: return nil
        }
    }

    private func emit(_ event: A11yEvent, coalescing eventClass: EventClass?, gate: @escaping () -> Bool) {
        let rendered = a11yRender(event: event)
        guard !rendered.text.isEmpty else { return }
        let priority = AnnouncementPriority(rendered.priority)
        if let eventClass {
            debounce(eventClass, text: rendered.text, priority: priority.nsPriority, gate: gate)
            return
        }
        // A High event — announced or relayed — supersedes every queued
        // line: navigation context is re-derivable by moving again, and a
        // stale count or value must not speak after the error (0a-9).
        if priority == .high {
            flushAllPending()
        }
        post(rendered.text, priority.nsPriority)
    }

    // MARK: Coalescing (mirrors CanvasAnnouncer)

    private func debounce(
        _ eventClass: EventClass, text: String, priority: NSAccessibilityPriorityLevel,
        gate: @escaping () -> Bool
    ) {
        pending[eventClass]?.work.cancel()
        // The class is `@MainActor` and every caller is therefore on the
        // main actor; this DispatchWorkItem is the one non-isolated closure
        // in the type and is dispatched ONLY to `DispatchQueue.main`, so
        // `assumeIsolated` makes that isolation explicit.
        let work = DispatchWorkItem { [weak self] in
            MainActor.assumeIsolated {
                guard let self else { return }
                guard let entry = self.pending.removeValue(forKey: eventClass) else { return }
                // Re-check relevance AT FIRE TIME.
                guard entry.gate() else { return }
                self.post(entry.text, entry.priority)
            }
        }
        pending[eventClass] = (work, text, priority, gate)
        DispatchQueue.main.asyncAfter(deadline: .now() + coalesceWindow, execute: work)
    }

    private func flushAllPending() {
        for (_, entry) in pending { entry.work.cancel() }
        pending = [:]
    }

    /// Drop any queued (debounced) announcements WITHOUT posting them —
    /// called when the graph view disappears or a vault opens/closes, so
    /// a coalesced count scheduled while the graph was active can't fire
    /// after the user has moved on.
    func cancelPending() {
        flushAllPending()
    }

    /// Test hook: emit pending debounced posts NOW, honoring each entry's
    /// fire-time gate (so gating is observable in tests, matching prod).
    func flushForTests() {
        let items = pending
        pending = [:]
        for (_, entry) in items {
            entry.work.cancel()
            guard entry.gate() else { continue }
            post(entry.text, entry.priority)
        }
    }
}

File: apps/slate-windows/src/SlateWindows/Graph/GraphAnnouncer.cs
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

using System.Diagnostics;
using System.Windows.Threading;
using uniffi.slate_uniffi;

namespace SlateWindows.Graph;

/// <summary>
/// W6-2 PR A (#746), contract A-10: the ONE announcement funnel for every
/// graph surface — the mac <c>GraphAnnouncer</c> twin and the sibling of
/// <c>CanvasAnnouncer</c>. The GRAMMAR is core's: callers hand over a typed
/// <c>GraphA11yEvent</c> from the PR 0a vocabulary; this class wraps it in
/// <c>A11yEvent.Graph</c>, renders it through the FFI, and posts the
/// rendered pair. What stays host-side is the 200 ms same-class
/// coalescing window and the High flush-and-drop (a11y.rs's pinned
/// list). No graph code posts on its own: the directory censuses parse
/// every file under <c>Graph/</c> and this is the one exempt file, by
/// its normalised path.
/// </summary>
internal sealed class GraphAnnouncer
{
    /// <summary>The coalescing classes core pins on the graph family
    /// (a11y.rs, "Coalescing class keys": navigation, filter,
    /// forceValue, settle). Membership is asserted from the Rust side by
    /// <c>the_windows_graph_coalescing_switch_matches_the_pinned_class_list</c>.</summary>
    private enum EventClass
    {
        Navigation,
        Filter,
        ForceValue,
        Settle,
    }

    /// <summary>The mac cadence: 200 ms latest-wins, per class.</summary>
    internal static readonly TimeSpan DefaultWindow = TimeSpan.FromMilliseconds(200);

    private readonly Action<RenderedAnnouncement> _post;
    private readonly Dictionary<EventClass, PendingLine> _pending = [];
    private readonly TimeSpan _window;
    private readonly Dispatcher _dispatcher;
    private bool _isShutDown;

    public GraphAnnouncer(Action<RenderedAnnouncement> post, TimeSpan? coalesceWindow = null)
    {
        ArgumentNullException.ThrowIfNull(post);
        _post = post;
        _window = coalesceWindow ?? DefaultWindow;
        _dispatcher = Dispatcher.CurrentDispatcher;
    }

    /// <summary>The only announcement API graph code may use.</summary>
    public void Announce(GraphA11yEvent @event)
    {
        ArgumentNullException.ThrowIfNull(@event);
        Emit(new A11yEvent.Graph(@event), CoalescingClassOf(@event), gate: null);
    }

    /// <summary>The filter count's GATED entry (0a-9's fire-time gate; A-10
    /// as amended, W6-2 PR B BD-12): the gate is stored with the pending
    /// line and re-checked when the window elapses — the mac's
    /// <c>graphTabActive</c> at fire (`GraphAnnouncer.swift:150`,
    /// `:194–207`) — so a count queued while the graph tab was effective
    /// is DROPPED if the tab left effective before it spoke. The census
    /// asserts the graph document enqueues its count through this entry
    /// and through no other.</summary>
    public void AnnounceGatedFilterCount(GraphA11yEvent.GraphFilterCount @event, Func<bool> gate)
    {
        ArgumentNullException.ThrowIfNull(@event);
        ArgumentNullException.ThrowIfNull(gate);
        Emit(new A11yEvent.Graph(@event), EventClass.Filter, gate);
    }

    /// <summary>Relay a core event that is NOT graph vocabulary — the
    /// shared <c>AccessibleDataGrid</c>'s sort, row-move and cell-move
    /// events — uncoalesced, with core's own priority.</summary>
    public void Relay(A11yEvent @event)
    {
        ArgumentNullException.ThrowIfNull(@event);
        Emit(@event, null, gate: null);
    }

    /// <summary>Render a graph event WITHOUT posting (contract A-6): the
    /// row's UIA Name and its row-move description are the same P1 copy
    /// the announcer would speak, obtained here so the relay stays the
    /// one file that renders.</summary>
    public static string RenderLabel(GraphA11yEvent @event)
    {
        ArgumentNullException.ThrowIfNull(@event);
        return SlateUniffiMethods.A11yRender(new A11yEvent.Graph(@event)).Text;
    }

    /// <summary>Retirement (contract A-1): every pending class dropped,
    /// nothing later accepted — the mac's <c>cancelPending</c>.</summary>
    internal void Shutdown()
    {
        _isShutDown = true;
        DropAllPending();
    }

    internal int RefusedAfterShutdownForTests { get; private set; }

    internal bool IsRetired => _isShutDown;

    private void Emit(A11yEvent @event, EventClass? eventClass, Func<bool>? gate)
    {
        if (_isShutDown)
        {
            RefusedAfterShutdownForTests++;
            Debug.Fail(
                "GraphAnnouncer posted after Shutdown: the workspace it belongs to has "
                + "been torn down, so nothing it says can be about a surface the user can see.");
            return;
        }
        if (!_dispatcher.CheckAccess())
        {
            _ = _dispatcher.BeginInvoke(() => Emit(@event, eventClass, gate));
            return;
        }
        RenderedAnnouncement rendered = SlateUniffiMethods.A11yRender(@event);
        if (rendered.Text.Length == 0)
        {
            Debug.Fail(
                $"the vocabulary rendered an empty string for {@event.GetType().Name}; "
                + "an announcement was dropped silently.");
            return;
        }
        if (eventClass is { } coalesced)
        {
            Debounce(coalesced, rendered, gate);
            return;
        }
        // a11y.rs: a High graph event, announced or relayed, flushes and
        // DROPS all four pending classes — never posts them.
        if (rendered.Priority == A11yPriority.High)
        {
            DropAllPending();
        }
        _post(rendered);
    }

    /// <summary>The coalescing class of a graph event, per the list
    /// pinned on the graph family core-side (0a-9): navigation for the
    /// row focus, filter for the count, forceValue for a force control,
    /// settle for the layout's settle; immediate for everything else.</summary>
    private static EventClass? CoalescingClassOf(GraphA11yEvent @event) => @event switch
    {
        GraphA11yEvent.GraphRow => EventClass.Navigation,
        GraphA11yEvent.GraphFilterCount => EventClass.Filter,
        GraphA11yEvent.GraphForceValue => EventClass.ForceValue,
        GraphA11yEvent.GraphLayoutSettled => EventClass.Settle,
        _ => null,
    };

    private void Debounce(EventClass eventClass, RenderedAnnouncement rendered, Func<bool>? gate)
    {
        if (!_pending.TryGetValue(eventClass, out PendingLine? line))
        {
            line = new PendingLine(_window, _dispatcher, () => Fire(eventClass));
            _pending[eventClass] = line;
        }
        line.Restart(rendered, gate);
    }

    private void Fire(EventClass eventClass)
    {
        if (!_pending.TryGetValue(eventClass, out PendingLine? line))
        {
            return;
        }
        (RenderedAnnouncement? rendered, Func<bool>? gate) = line.Take();
        if (rendered is null)
        {
            return;
        }
        // 0a-9's fire-time gate (A-10 as amended): a stored gate that has
        // gone false since the line was queued DROPS it.
        if (gate is not null && !gate())
        {
            DroppedAtFireForTests++;
            return;
        }
        _post(rendered);
    }

    /// <summary>How many gated lines the fire-time gate dropped.</summary>
    internal int DroppedAtFireForTests { get; private set; }

    /// <summary>Drop every pending class without posting — the mac's
    /// <c>cancelPending</c> on view departure (0a-9); a document's
    /// retirement calls it on the shared relay (A-1 as amended).</summary>
    internal void DropAllPending()
    {
        foreach (PendingLine line in _pending.Values)
        {
            _ = line.Take();
        }
    }

    /// <summary>Test hook: emit every pending debounced line NOW (the
    /// mac <c>flushForTests</c> twin).</summary>
    internal void FlushForTests()
    {
        foreach (EventClass eventClass in _pending.Keys.ToList())
        {
            Fire(eventClass);
        }
    }

    /// <summary>How many classes hold a pending line — the retirement
    /// fact's witness.</summary>
    internal int PendingForTests => _pending.Values.Count(line => line.HasLine);

    /// <summary>One class's queued line and its timer; the timer stops
    /// whenever the line is taken.</summary>
    private sealed class PendingLine
    {
        private readonly DispatcherTimer _timer;
        private RenderedAnnouncement? _rendered;
        private Func<bool>? _gate;

        internal PendingLine(TimeSpan window, Dispatcher dispatcher, Action onElapsed)
        {
            _timer = new DispatcherTimer(window, DispatcherPriority.Normal, (_, _) => onElapsed(), dispatcher);
            _timer.Stop();
        }

        internal bool HasLine => _rendered is not null;

        /// <summary>Latest wins — the line AND its gate (0a-9: the gate is
        /// stored with the pending line, re-checked at fire).</summary>
        internal void Restart(RenderedAnnouncement rendered, Func<bool>? gate)
        {
            _rendered = rendered;
            _gate = gate;
            _timer.Stop();
            _timer.Start();
        }

        internal (RenderedAnnouncement? Rendered, Func<bool>? Gate) Take()
        {
            _timer.Stop();
            RenderedAnnouncement? rendered = _rendered;
            Func<bool>? gate = _gate;
            _rendered = null;
            _gate = null;
            return (rendered, gate);
        }
    }
}
